numpy/typing/tests/data/pass/shape.py

```diff
@@ -0,0 +1,28 @@
+from typing import Any, cast, Literal, NewType, TypeVar
+
+import numpy as np
+from typing_extensions import assert_type
+
+Time = NewType("Time", int)
+Series = NewType("Series", int)
+
+# Test bound
+arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
+
+Ax1 = TypeVar("Ax1", bound=int)
+Ax2 = TypeVar("Ax2", bound=int)
+def transpose(
+    a: np.ndarray[tuple[Ax1, Ax2], Any]
+) -> np.ndarray[tuple[Ax2, Ax1], Any]:
+    return cast(np.ndarray[tuple[Ax2, Ax1], Any], np.transpose(a))
+
+# Test variance
+assert_type(transpose(arr).shape, tuple[Series, Time])
+
+# Test shape generic
+assert_type(arr.shape, tuple[Time, Series])
+
+# Test literal
+arr2: np.ndarray[tuple[Literal[1], Literal[2]], Any] = np.arange(2).reshape((1, 2))
+assert_type(arr2.shape, tuple[Literal[1], Literal[2]])
+assert_type(transpose(arr2).shape, tuple[Literal[2], Literal[1]])
```

Comment:
This also tests (co)variance of the tuple values like the test on line 20. So one of them is redundant.

Reply:
There are tests for the covariance of the tuple *values*, but not yet for the covariance of the tuple *itself*. So this doesn't actually test the covariance of the shape type. 
So e.g. passing some `typing.NamedTuple` instance as a shape.

BTW, you might wanna check whether this'll match with the runtime behavior (I don't mean as a test, just in a python shell or something)

Reply:
Oooh, interesting point, and you're right.  I was less familiar with using `Literal`, but apparently it is treated as a subtype in a way "identical to how we handle NewTypes."

I think both do handle covariance of `tuple` itself, because tuple is covariant with respect to _its_ arguments - `tuple[Time, Literal[2]]` is a subtype of `tuple`.  But you're right it makes sense to add a test using a real tuple subclass.

Reply:
If you meant that `reshape` works with a namedtuple of ints, then yeah, just checked it :heavy_check_mark:

Reply:
> On the last part, I'm not sure what you mean. What runtime behavior is uncertain?

```pycon
>>> import numpy as np
>>> from typing import NamedTuple
>>> class Shape2D(NamedTuple):
...     x: int
...     y: int
... 
>>> x = np.empty(Shape2D(6, 7))
>>> type(x.shape)
<class 'tuple'>
```

Reply:
Ah! It would be cool.  Alas

Reply:
I've opened an issue about it: https://github.com/numpy/numpy/issues/27086